### PR TITLE
Generator template api

### PR DESCRIPTION
**Pull request by jgraff2**

### CODE REVIEW

**Comment by jgraff2** [comment]
The tests won't pass until the swagger changes are merged. Waiting to add reviewers.

**Comment by jgraff2** [comment]
db/model/generatortemplate.js, line 68:
 // TODO: revisit toUrl validation, while doing the api changes

The validation for making sure 'url' and 'toUrl' are not both present or both absent works fine, and the validation of the actual toUrl function is done in the collector. Is there something else that needs to be done?

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  let user;
+  const generatorTemplateToCreate = u.getGeneratorTemplate();
+
+  beforeEach((done) => {
+    tu.createToken()
+    .then((returnedToken) => {
+      token = returnedToken;
+      done();
+    })
+    .catch((err) => done(err));
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: .catch(done);

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  });
+
+  beforeEach((done) => {
+    GeneratorTemplate.create(generatorTemplateToCreate)
+    .then((gen) => {
+      generatorTemplate = gen;
+    }).then(() =>
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: pls. see one of the updated tests from this weekend to see the formatting and structure of this boilerplate-ish beforeeach block

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+      generatorTemplate.addWriter(secUsr);
+      user = secUsr;
+    })
+    .then(() => tu.createThirdUser())
+    .then((tUsr) => tu.createTokenFromUserName(tUsr.name))
+    .then((tkn) => {
+      otherValidToken = tkn;
+    })
+    .then(() => done())
+    .catch((err) => done(err));
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: .catch(done);

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+          return done(err);
+        }
+
+        api.get(writersPath.replace('{key}', generatorTemplate.id))
+        .set('Authorization', token)
+        .expect(constants.httpStatus.OK)
+        .expect((res) => {
+          expect(res.body).to.have.length(1);
+        })
+        .end((_err /* , res */) => {
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: replace lines 87-94 with just `.end(done);` -- ask me if you don't think it's the same thing

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: same comment as above

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+          return done(err);
+        }
+
+        api.get(writersPath.replace('{key}', generatorTemplate.id))
+        .set('Authorization', token)
+        .expect(constants.httpStatus.OK)
+        .expect((res) => {
+          expect(res.body).to.have.length(2);
+        })
+        .end((_err /* , res */) => {
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: same comment as above

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      });
+    });
+
+    it('return 403 when deleteting a writer using a token ' +
+      'generated for a user not already in the list of writers', (done) => {
+      api.delete(writerPath.replace('{key}', generatorTemplate.id)
+      .replace('{userNameOrId}', 'invalidUserName'))
+      .set('Authorization', otherValidToken)
+      .expect(constants.httpStatus.FORBIDDEN)
+      .end((err /* , res */) => {
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: same comment as above

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * tests/api/v1/generatorTemplates/deleteWriters.js
+ */
+'use strict';
+
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: no empty line here

File: tests/api/v1/generatorTemplates/get.js [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * tests/api/v1/generators/get.js
+ */
+'use strict';
+
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: no empty line here

File: tests/api/v1/generatorTemplates/get.js [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+  after(u.forceDelete);
+  after(tu.forceDeleteUser);
+
+  it('simple GET OK', (done) => {
+    api.get(`${path}`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+  });
+
+  it('using ?sort=name sort the results in alphabetical order of ' +
+    'generator name', (done) => {
+    api.get(`${path}?sort=name`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+  });
+
+  it('using ?sort=-name should sort the results in reverse alphabetical ' +
+    'order of generator name', (done) => {
+    api.get(`${path}?sort=-name`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+    });
+  });
+
+  it('Simple GET with id', (done) => {
+    api.get(`${path}/${template1.id}`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+    });
+  });
+
+  it('Simple GET with name', (done) => {
+    api.get(`${path}/${template1.name}`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+    });
+  });
+
+  it('Simple GET with name in lowercase', (done) => {
+    api.get(`${path}/${template1.name.toLowerCase()}`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 16]
```diff
@@ -0,0 +1,372 @@
...
+    });
+  });
+
+  it('find by tags, found', (done) => {
+    api.get(`${path}?tags=tag1`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 17]
```diff
@@ -0,0 +1,372 @@
...
+    });
+  });
+
+  it('find by tags multiple, found', (done) => {
+    api.get(`${path}?tags=tag1,tag2`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 18]
```diff
@@ -0,0 +1,372 @@
...
+    });
+  });
+
+  it('find by tags, not found', (done) => {
+    api.get(`${path}?tags=tag5`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(err);

File: tests/api/v1/generatorTemplates/get.js [hunk 19]
```diff
@@ -0,0 +1,372 @@
...
+      }
+
+      expect(res.body).to.have.length(ZERO);
+      done();
+    });
+  });
+
+
+  describe('get with fields - ', () => {
+
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: no empty line here

File: tests/api/v1/generatorTemplates/get.js [hunk 20]
```diff
@@ -0,0 +1,372 @@
...
+      const extraFields = ['apiLinks'];
+      if (!fields.includes('id')) extraFields.push('id');
+      const expectedFields = [...extraFields, ...fields];
+      api.get(`${path}?fields=${fields.join(',')}`)
+      .set('Authorization', token)
+      .expect(constants.httpStatus.OK)
+      .end((err, res) => {
+        if (err) done(err);
+        if (res.body.errors) {
+          done(res.body.errors[0]);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: return done(...) here

File: tests/api/v1/generatorTemplates/get.js [hunk 21]
```diff
@@ -0,0 +1,372 @@
...
+
+    function getAllWithFields(done, ...fields) {
+      const extraFields = ['apiLinks'];
+      if (!fields.includes('id')) extraFields.push('id');
+      const expectedFields = [...extraFields, ...fields];
+      api.get(`${path}?fields=${fields.join(',')}`)
+      .set('Authorization', token)
+      .expect(constants.httpStatus.OK)
+      .end((err, res) => {
+        if (err) done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]:
```
if (err) {
  return done(err);
}

```

File: tests/api/v1/generatorTemplates/get.js [hunk 22]
```diff
@@ -0,0 +1,372 @@
...
+        if (res.body.errors) {
+          done(res.body.errors[0]);
+        }
+        expect(res.body).to.have.length(4);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: empty line after if block

File: tests/api/v1/generatorTemplates/get.js [hunk 23]
```diff
@@ -0,0 +1,372 @@
...
+
+    function getByKeyWithFields(done, ...fields) {
+      const extraFields = ['apiLinks'];
+      if (!fields.includes('id')) extraFields.push('id');
+      const expectedFields = [...extraFields, ...fields];
+      api.get(`${path}/${template1.id}?fields=${fields.join(',')}`)
+      .set('Authorization', token)
+      .expect(constants.httpStatus.OK)
+      .end((err, res) => {
+        if (err) done(err);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: formatting and return done

File: tests/api/v1/generatorTemplates/get.js [hunk 24]
```diff
@@ -0,0 +1,372 @@
...
+      getByKeyWithFields(done, 'description', 'name', 'tags');
+    });
+
+    it(`fields param by key - all fields`, (done) => {
+      getByKeyWithFields(done, ...fields);
+    });
+  });
+
+  describe('find by fields:', () => {
+
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: apply all comments from above to remaining

File: tests/api/v1/generatorTemplates/get.js [hunk 25]
```diff
@@ -0,0 +1,372 @@
...
+    });
+
+    it('find by isPublished non-boolean (error)', (done) => {
+      findByField(done, 'isPublished', 'aaa', 'err');
+    });
+
+    it('find by isPublished wildcard (error)', (done) => {
+      findByField(done, 'isPublished', 'f*lse', 'err');
+    });
+
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: no empty line here

File: tests/api/v1/generatorTemplates/get.js [hunk 26]
```diff
@@ -0,0 +1,372 @@
...
+  });
+
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: no empty line here

File: tests/api/v1/generatorTemplates/getWriters.js [hunk 27]
```diff
@@ -0,0 +1,166 @@
+/**
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/getWriters.js]: apply all comments from other tests here too

File: tests/api/v1/generatorTemplates/patch.js [hunk 28]
```diff
@@ -0,0 +1,120 @@
+/**
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/patch.js]: apply all comments from other tests here too

File: tests/api/v1/generatorTemplates/post.js [hunk 29]
```diff
@@ -0,0 +1,120 @@
+/**
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/post.js]: apply all comments from other tests here too

File: tests/api/v1/generatorTemplates/postWithCreatedBy.js [hunk 30]
```diff
@@ -0,0 +1,130 @@
+/**
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/postWithCreatedBy.js]: apply all comments from other tests here too

File: tests/api/v1/generatorTemplates/postWriters.js [hunk 31]
```diff
@@ -0,0 +1,124 @@
+/**
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/postWriters.js]: apply all comments from other tests here too

File: tests/api/v1/generatorTemplates/put.js [hunk 32]
```diff
@@ -0,0 +1,119 @@
+/**
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/put.js]: apply all comments from other tests here too

File: tests/api/v1/generatorTemplates/putPatchWithoutPerms.js [hunk 33]
```diff
@@ -0,0 +1,113 @@
+/**
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/putPatchWithoutPerms.js]: apply all comments from other tests here too

File: tests/db/model/generatortemplate/utils.js [hunk 34]
```diff
@@ -17,7 +17,7 @@ const testStartTime = new Date();
 const GT_SIMPLE = {
   name: 'refocus-ok-template',
   description: 'Collect status data',
-  keywords: [
+  tags: [
```
Comment (iamigo) [inline tests/db/model/generatortemplate/utils.js]: careful about this change and the timing of the other PR

File: db/model/generatortemplate.js [hunk 35]
```diff
@@ -120,11 +129,7 @@ module.exports = function user(seq, dataTypes) {
         is: constants.versionRegex,
       },
     },
-    bulk: {
-      type: dataTypes.BOOLEAN,
```
Comment (shriramshankar) [inline db/model/generatortemplate.js]: deleting a column also requires a migration.

File: tests/api/v1/generatorTemplates/get.js [hunk 36]
```diff
@@ -0,0 +1,372 @@
...
+ * Copyright (c) 2017, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or
+ * https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * tests/api/v1/generators/get.js
+ */
```
Comment (shriramshankar) [inline tests/api/v1/generatorTemplates/get.js]: generatorTemplate

File: db/model/generatortemplate.js (hunk 35 above)
Comment (jgraff2) [inline db/model/generatortemplate.js]: ok

File: tests/api/v1/generatorTemplates/get.js [hunk 37]
```diff
@@ -0,0 +1,372 @@
...
+const u = require('./utils');
+const GeneratorTemplate = tu.db.GeneratorTemplate;
+const path = '/v1/generatorTemplates';
+const expect = require('chai').expect;
+const ZERO = 0;
+const ONE = 1;
+const TWO = 2;
+const THREE = 3;
+
+describe(`api: GET ${path}`, () => {
```
Comment (shriramshankar) [inline tests/api/v1/generatorTemplates/get.js]: This one just came out of the review in #525 .------ "the outermost "describe" for every test should be named for the file it's in (and every "describe" name should end with " >")" . We can choose to correct it now or later.

Comment (iamigo) [inline tests/api/v1/generatorTemplates/get.js]: later's fine

File: db/model/generatortemplate.js (hunk 35 above)
Comment (jgraff2) [inline db/model/generatortemplate.js]: can you explain the naming convention for migrations?

Comment (iamigo) [inline db/model/generatortemplate.js]: https://salesforce.github.io/refocus/docs/40-coredevelopment.html

the migration tool generates the file (named based on timestamp), you just give it some short descriptive name which gets appended to the filename it generates

**Comment by jgraff2** [comment]
@shriramshankar can you comment on the question about the TODO above?

**Comment by jgraff2** [comment]
migrations added.

**Comment by shriramshankar** [comment]
@jgraff2 Sorry I missed that. As long as it works we are good. Thanks.

File: migrations/20170816230215-generatorTemplateRemoveBulk.js [hunk 38]
```diff
@@ -0,0 +1,35 @@
...
+     Example:
+     return qi.dropTable('users');
+    */
+    return qi.addColumn('GeneratorTemplates', 'bulk', {
+      type: Sequelize.BOOLEAN,
+      defaultValue: false,
+    });
+  },
+};
+
```
Comment (shriramshankar) [inline migrations/20170816230215-generatorTemplateRemoveBulk.js]: You can combine these two into a single migration file. The promises can be chained and you might want to wrap it in a transaction. Something like this on here 
https://github.com/salesforce/refocus/blob/master/migrations/20170330184249-dropSampleColumnHardDelete.js

File: migrations/20170816230206-generatorTemplateChangeKeywordsToTags.js [hunk 39]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Copyright (c) 2016, salesforce.com, inc.
```
Comment (iamigo) [inline migrations/20170816230206-generatorTemplateChangeKeywordsToTags.js]: This one is from the other PR, right?

File: migrations/20170816230215-generatorTemplateRemoveBulk.js [hunk 40]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright (c) 2016, salesforce.com, inc.
```
Comment (iamigo) [inline migrations/20170816230215-generatorTemplateRemoveBulk.js]: 2017

File: migrations/20170816230215-generatorTemplateRemoveBulk.js (hunk 38 above)
Comment (jgraff2) [inline migrations/20170816230215-generatorTemplateRemoveBulk.js]: I was doing that originally but I thought I should separate them since the two changes are unrelated. Should I move them back to a single file?

File: migrations/20170816230206-generatorTemplateChangeKeywordsToTags.js (hunk 39 above)
Comment (jgraff2) [inline migrations/20170816230206-generatorTemplateChangeKeywordsToTags.js]: I'm making the same change to both Generators and GeneratorTemplates. This PR is for GeneratorTemplates, including both api and model changes. The other PR is for the Generators model change.

**Comment by jgraff2** [comment]
removed TODO.

File: migrations/20170816230215-generatorTemplateRemoveBulk.js (hunk 38 above)
Comment (shriramshankar) [inline migrations/20170816230215-generatorTemplateRemoveBulk.js]: It is nice to have them in a single file since both belong to the same table and both are changing the table definition. But don't worry about this change unless you really want to; approved this PR already.

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 41]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        api.get(writersPath.replace('{key}', generatorTemplate.id))
+        .set('Authorization', token)
+        .expect(constants.httpStatus.OK)
+        .expect((res) => {
+          expect(res.body).to.have.length(1);
+        })
+        .end(done);
+        return null;
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: no return null here -- line 85 returns done() if there's no error

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 42]
```diff
@@ -0,0 +1,142 @@
...
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: same comment as above

File: tests/api/v1/generatorTemplates/deleteWriters.js [hunk 43]
```diff
@@ -0,0 +1,142 @@
...
+        }
+
+        api.get(writersPath.replace('{key}', generatorTemplate.id))
+        .set('Authorization', token)
+        .expect(constants.httpStatus.OK)
+        .expect((res) => {
+          expect(res.body).to.have.length(2);
+        })
+        .end(done);
+        return null;
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/deleteWriters.js]: same again

File: tests/api/v1/generatorTemplates/post.js [hunk 44]
```diff
@@ -0,0 +1,119 @@
...
+  it('simple post OK', (done) => {
+    api.post(path)
+    .set('Authorization', token)
+    .send(generatorTemplate)
+    .expect(constants.httpStatus.CREATED)
+    .end((err, res) => {
+      if (err) {
+        return done(err);
+      }
+      expect(res.body.apiLinks).to.be.an('Array');
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/post.js]: this isn't going to pass jscs check - needs blank line after if block

File: tests/api/v1/generatorTemplates/post.js [hunk 45]
```diff
@@ -0,0 +1,119 @@
...
+      .set('Authorization', token)
+      .send(generatorTemplate)
+      .expect(constants.httpStatus.FORBIDDEN)
+      .end((err, res) => {
+        if (err) {
+          return done(err);
+        }
+
+        expect(res.body.errors[ZERO].type)
+          .to.equal(tu.uniErrorName);
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/post.js]: when an expect(...) like this all fits on one line in under 80 chars, don't break it into two lines

File: tests/api/v1/generatorTemplates/putPatchWithoutPerms.js [hunk 46]
```diff
@@ -0,0 +1,110 @@
...
+
+  it('PUT without permission: should return 403', (done) => {
+    const toPut = {
+      name: 'template1',
+      description: 'this is template1...',
+      tags: [
+        'tag1',
+        'tag2'
+      ],
+      "author": {
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/putPatchWithoutPerms.js]: no quotes around attribute names - this will fail jscs check

File: tests/api/v1/generatorTemplates/putPatchWithoutPerms.js [hunk 47]
```diff
@@ -0,0 +1,110 @@
...
+        "name": "author1",
```
Comment (iamigo) [inline tests/api/v1/generatorTemplates/putPatchWithoutPerms.js]: and no double quotes around attribute values - we use single quotes - this will also fail jscs check
